packages/xo-web/src/xo-app/vm/tab-network.js

```diff
@@ -527,12 +538,23 @@ export default class TabNetwork extends BaseComponent {
       newVif: !this.state.newVif,
     })
 
-  _getIpAddresses = createSelector(
+  _getIpsByDevice = createSelector(
     () => this.props.vm.addresses,
-    // VM_guest_metrics.networks seems to always have 3 fields (ip, ipv4 and ipv6) for each interface
-    // http://xenbits.xenproject.org/docs/4.12-testing/misc/xenstore-paths.html#attrvifdevidipv4index-ipv4_address-w
-    // https://github.com/xapi-project/xen-api/blob/d650621ba7b64a82aeb77deca787acb059636eaf/ocaml/xapi/xapi_guest_agent.ml#L76-L79
-    addresses => uniq(values(addresses))
+    addresses => {
+      // VM_guest_metrics.networks seems to always have 3 fields (ip, ipv4 and ipv6) for each interface
+      // http://xenbits.xenproject.org/docs/4.12-testing/misc/xenstore-paths.html#attrvifdevidipv4index-ipv4_address-w
+      // https://github.com/xapi-project/xen-api/blob/d650621ba7b64a82aeb77deca787acb059636eaf/ocaml/xapi/xapi_guest_agent.ml#L76-L79
+      const ipsByDevice = {}
+      Object.keys(addresses).forEach(key => {
+        const device = key.split('/')[0]
+        const ips = ipsByDevice[device] || []
+        const ip = addresses[key]
+        if (!ips.includes(ip)) {
+          ipsByDevice[device] = ips.concat(ip)
```

Comment:
On each iteration, we create a new array containing the old values with the new value. We can improve that by creating one time an array, then, push the new value if doesn't exists.

```js
 Object.keys(addresses).forEach(key => {
  const device = key.split('/')[0]
  let ips = ipsByDevice[device]
  if (ips === undefined) {
   ips = ipsByDevice[device] = []
  }

  const ip = addresses[key]
  if (!ips.includes(ip)){
    ips.push(ip)
  }
})
```

Reply:
I don't think we even need to check if the address is already in the array. If the hypervisor reports the same IP twice for the same device (I don't know if it's possible), then we can report it as well.

Reply:
We can have a duplicated IP due to a compatibility of different protocols `n/ip` entry is the same than `n/ipv4/0` entry. See https://github.com/xapi-project/xen-api/blob/324bc6ee6664dd915c0bbe57185f1d6243d9ed7e/ocaml/xapi/xapi_guest_agent.ml#L59-L81